rust/src/quic/frames.rs

```diff
@@ -0,0 +1,236 @@
+    Irtt = 0x49525454,
+    Kexs = 0x4b455853,
+    Mids = 0x4d494453,
+    Mspc = 0x4d535043,
+    Nonc = 0x4e4f4e43,
+    Nonp = 0x4e4f4e50,
+    Pad = 0x50414400,
+    Pdmd = 0x50444d44,
+    Pubs = 0x50554253,
+    Scid = 0x53434944,
+    Scls = 0x53434c53,
+    Sfcw = 0x53464357,
+    Smhl = 0x534d484c,
+    Sni = 0x534e4900,
+    Sno = 0x534e4f00,
+    Stk = 0x53544b00,
+    Tcid = 0x54434944,
+    Uaid = 0x55414944,
+    Ver = 0x56455200,
+    Xlct = 0x584c4354,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum Frame {
+    Padding,
+    Stream(Stream),
+    Unknown(Vec<u8>),
+}
+
+fn parse_tag(input: &[u8]) -> Result<(&[u8], StreamTag), QuicError> {
+    let (rest, tag) = be_u32(input)?;
+
+    let tag = StreamTag::from_u32(tag)
+        .ok_or(QuicError::Parse(format!("Cannot match StreamTag: {}", tag)))?;
+
+    Ok((rest, tag))
+}
+
+fn parse_tag_and_offset(input: &[u8]) -> Result<(&[u8], TagOffset), QuicError> {
+    let (rest, tag) = parse_tag(input)?;
+    let (rest, offset) = le_u32(rest)?;
+
+    Ok((rest, (tag, offset)))
+}
+
+fn parse_crypto_stream(input: &[u8]) -> Result<(&[u8], Vec<TagValue>), QuicError> {
+    // [message tag][number of tag entries: N][pad][[tag][end offset], ...N][value data]
+    let (rest, _message_tag) = parse_tag(input)?;
+
+    let (rest, num_entries) = le_u16(rest)?;
+    let (rest, _padding) = take(2usize)(rest)?;
+
+    let mut tags_offset = Vec::new();
+    let rest = {
+        let mut rest = rest;
+
+        for _i in 0..num_entries {
+            let (new_rest, (tag, offset)) = parse_tag_and_offset(rest)?;
+            rest = new_rest;
+
+            tags_offset.push((tag, offset));
+        }
+
+        rest
+    };
+
+    // Convert (Tag, Offset) to (Tag, Value)
+    let mut tags = Vec::new();
+    let mut previous_offset = 0;
+    let mut rest = rest;
+    for (tag, offset) in tags_offset {
+        let offset = usize::try_from(offset).map_err(|e| {
+            QuicError::Parse(format!(
+                "Could not convert offset to usize: {}",
+                e.to_string()
+            ))
+        })?;
+
+        if offset < previous_offset {
+            // offsets should be increasing
+            SCLogDebug!(
+                "Invalid packet, offsets not increasing: offset {} previous {}",
+                offset,
+                previous_offset
+            );
+            return Err(QuicError::InvalidPacket);
+        }
+
+        let value_len = offset - previous_offset;
+        let (new_rest, value) = take(value_len)(rest)?;
+
+        previous_offset = offset;
+        rest = new_rest;
+
+        tags.push((tag, value.to_vec()))
+    }
+
+    Ok((rest, tags))
+}
+
+fn parse_stream_frame(input: &[u8], frame_ty: u8) -> Result<(&[u8], Frame), QuicError> {
+    let rest = input;
+
+    // 0b1_f_d_ooo_ss
+    let fin = frame_ty & 0x40 == 0x40;
+    let has_data_length = frame_ty & 0x20 == 0x20;
+
+    let offset_hdr_length = {
+        let mut offset_length = (frame_ty & 0x1c) >> 2;
```

Comment:
This looks complex 
 Could we replace it with 
`let offset_hdr_length = if frame_ty < 4 { 0 } else { 1 + (frame_ty & 0x1c) >> 2 } ;`